source/opt/fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,565 @@
+            "means its nested values must be scalar type constants");
+        concatenated_id_vec.push_back(nested_const_id);
+      }
+    } else {
+      // Does not find the operand in the id_to_const_val_ map, the
+      // value of the operand might not be fully resolved yet, so this
+      // OpSpecConstantOp instruction can not be folded.
+      return nullptr;
+    }
+  }
+  auto result = std::unique_ptr<Value>(new CompositeValue);
+  for (uint32_t i = 3; i < inst->NumInOperands(); i++) {
+    assert(inst->GetInOperand(i).type == SPV_OPERAND_TYPE_LITERAL_INTEGER &&
+           "The literal operand must have type "
+           "SPV_OPERAND_TYPE_LITERAL_INTEGER");
+    uint32_t literal = inst->GetSingleWordInOperand(i);
+    assert(literal < concatenated_id_vec.size() &&
+           "Literal index out of bound of the concatenated vector");
+
+    result->AsComposite()->nested.push_back(concatenated_id_vec[literal]);
+  }
+  return result;
+}
+
+std::tuple<std::unique_ptr<Value>, std::vector<uint32_t>>
+FoldSpecConstantOpAndCompositePass::GetBroadcastOperationResult(
+    ir::Instruction* inst) {
+  const uint32_t spec_opcode_count = 1;
+  uint32_t num_operands = inst->NumInOperands();
+  uint32_t operand_index = 1;
+  uint32_t operand_id = inst->GetSingleWordInOperand(operand_index);
+  auto iter = id_to_const_val_.find(operand_id);
+  if (iter == id_to_const_val_.end()) {
+    return std::make_tuple(nullptr, std::vector<uint32_t>());
+
+  } else if (iter->second->AsScalar()) {
+    // Get the value in raw words for all the operands, prepare for the
+    // operation.
+    std::vector<uint32_t> operand_values_in_raw_words;
+    operand_values_in_raw_words.reserve(num_operands - spec_opcode_count);
+    operand_values_in_raw_words.push_back(iter->second->AsScalar()->word);
+    for (uint32_t i = operand_index + 1; i < num_operands; i++) {
+      iter = id_to_const_val_.find(inst->GetSingleWordInOperand(i));
+      if (iter == id_to_const_val_.end())
+        return std::make_tuple(nullptr, std::vector<uint32_t>());
+      // All the operands should be scalar constants.
+      assert(iter->second->AsScalar() && "All operands must be in scalar type");
+      operand_values_in_raw_words.push_back(iter->second->AsScalar()->word);
+    }
+    uint32_t result_in_raw_word =
+        Operate(static_cast<SpvOp>(inst->GetSingleWordInOperand(0)),
+                std::move(operand_values_in_raw_words));
+    return std::make_tuple(
+        std::unique_ptr<Value>(new ScalarValue(result_in_raw_word)),
+        std::vector<uint32_t>());
+
+  } else {
+    assert(
+        AllInOperandValuesAreConstVectorsInSameLength(inst, id_to_const_val_) &&
+        "All the operands must be vector constants with same number of "
+        "components");
+    (void)static_cast<bool (*)(
+        ir::Instruction*,
+        const std::unordered_map<uint32_t, std::unique_ptr<Value>>&)>(
+        AllInOperandValuesAreConstVectorsInSameLength);
```

Comment:
This is to fix an compilation error in VS.

If we use `(void)AllInOperandValuesAreConstVectorsInSameLength` here, VS will consider it as an function call and complain the argument list is missing.